### ISSUE TITLE
Fails to download agent when building with mix_docker

### ISSUE THREAD

**Issue by jeremytregunna** [appsignal/appsignal-elixir#213]
I use [mix_docker](https://github.com/Recruitee/mix_docker) to build, release and deploy docker images. As part of it's phase, it generates two docker files, one for building, one for releasing. It's in this first phase, building, when I compile dependencies, that it fails to download the agent because of a missing directory (presumably /tmp, even though docker complains if I try and create it that it already exists).

It's unclear to me what can be done to resolve this case, below is the output as taken from mix_docker's build process. None of the other issues with docker seem to fit exactly with what I'm experiencing. Any help would be appreciated.

```
19:24:20.303 [info]  Downloading agent release from https://appsignal-agent-releases.global.ssl.fastly.net/7ade8ff/appsignal-x86_64-linux-all-static.tar.gz
==> appsignal
could not compile dependency :appsignal, "mix compile" failed. You can recompile this dependency with "mix deps.compile appsignal", update it with "mix deps.update appsignal" or clean it with "mix deps.clean appsignal"
** (ErlangError) erlang error: :enoent
    (elixir) lib/system.ex:564: System.cmd("curl", ["-s", "-S", "--retry", "5", "-f", "-o", "/tmp/appsignal-agent-7ade8ff.tar.gz", "https://appsignal-agent-releases.global.ssl.fastly.net/7ade8ff/appsignal-x86_64-linux-all-static.tar.gz"], [stderr_to_stdout: true])
    mix_helpers.exs:64: Mix.Appsignal.Helper.download_file/2
    mix_helpers.exs:52: Mix.Appsignal.Helper.download_and_extract/3
    mix_helpers.exs:35: Mix.Appsignal.Helper.ensure_downloaded/1
    /opt/app/deps/appsignal/mix.exs:13: Mix.Tasks.Compile.Appsignal.run/1
    (mix) lib/mix/task.ex:294: Mix.Task.run_task/3
    (elixir) lib/enum.ex:1229: Enum."-map/2-lists^map/1-0-"/2
    (mix) lib/mix/tasks/compile.all.ex:19: anonymous fn/1 in Mix.Tasks.Compile.All.run/1

The command '/bin/sh -c mix do deps.get, deps.compile' returned a non-zero code: 1
** (MatchError) no match of right hand side value: {%IO.Stream{device: :standard_io, line_or_bytes: :line, raw: false}, 1}
    lib/mix_docker.ex:180: MixDocker.system!/2
    lib/mix_docker.ex:20: MixDocker.build/1
    (mix) lib/mix/task.ex:294: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:58: Mix.CLI.run_task/2
    (elixir) lib/code.ex:370: Code.require_file/2
```

**Comment by tombruijn** [comment]
Hi Jeremy,

We looked into this, but didn't immediately think of a solution to fix it quickly. We'll need to have a look at why we're using the system's tmp directory for this. I don't think we have to.

For now you can have a look at what the ownership and permissions are for the tmp directory. They should be the writable by the user that's installing the Elixir package.

Let me know if that's the case or not

-- continued on intercom

**Comment by jeremytregunna** [comment]
Hi @tombruijn, thanks for the reply. I have a workaround that is functioning, but the problem still persists. My workaround involves using the `ADD` command to instruct docker to download and place the archive in /tmp/ as app signal's mix helper expects. Obviously this is inherently flawed as appsignal-elixir version updates would break this, so for now this is the solution I'm using and locking my version of appsignal-elixir down.

More to your question, /tmp is set 1777 as is normal, so I'm not sure that's the issue.

**Comment by tombruijn** [comment]
@jeremytregunna okay, so you download the agent during building the docker image, that's probably as the root user. Is your app also started under the root user? Does the downloaded file also have the 1777 permissions?

**Comment by jeremytregunna** [comment]
This all happens as the root user, yes. Downloaded file has 644 permissions.

**Comment by tombruijn** [comment]
Hmm.. That's odd. But is it working for you right now or can you not complete the installation?

> I have a workaround that is functioning, but the problem still persists.

confused me a bit

**Comment by jeremytregunna** [comment]
@tombruijn What I'm saying is the problem with the dependency installation still exists inside the docker container for me. My workaround means that due to the way appsignal-elixir checks if it's already got the file, I can have docker download the image, fix my dependency at a specific version, and complete an install (so appsignal-elixir doesn't do any downloading of anything, and I get the files added to the image before deps.compile runs). That's all. The issue is still a very real one, but I've been able to continue in spite of it.

**Comment by tombruijn** [comment]
@jeremytregunna could you try out the branch from PR #216 and let me know if it works?

Should move the download directory to the package's private path which should have the appropriate permissions. Then you no longer need your workaround.

**Comment by tombruijn** [comment]
Merged an attempted fix in #231. But since I cannot reproduce I'm not 100% sure.
